docs/spec_draft.md

```diff
@@ -358,6 +359,56 @@ IEEE-754 specification.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.concatenate
+
+### Semantics
+
+Concatenates sequence of tensors in `val` along dimension `dimension` and
+produces a tensor as a result. The tensors are of the same rank as the input
+operands except the specified dimension.
+
+### Operands
+
+| Name        | Type                                                                     |
+|-------------|--------------------------------------------------------------------------|
+| `dimension` | `i64`                                                                    |
+| `val`       | Sequence of tensors of integer, floating-point, or complex element types |
+
+### Results
+
+| Name     | Type                                                        |
+|----------|-------------------------------------------------------------|
+| `result` | tensor of integer, floating-point, or complex element types |
+
+### Constraints
+
+  * (C1) `val` and `result` have the same type.
+  * (C2) All tensors in `val` have the same element type and rank.
+  * (C3) Tensors of `val` all have the same dimension except for `dimension`.
```

Comment:
I have a few feedback items, but I think they'll be easier to share via a proposed rewording:

  * (C1) All tensors in `val` have the same element type.
  * (C2) All tensors in `val` have the same shape except for the size of the `dimension` dimension.
  * (C3) `dimension` is non-negative and less than the rank of the tensors in `val`.
  * (C4) `result` has the same element type as the tensors in `val`.
  * (C5) `result` has the same shape as the tensors in `val` except for the size of the `dimension` dimension, which is calculated as a sum of the sizes of this dimension in all tensors in `val`.

Feedback items:
  1. Let's go from left to right, i.e. first `val`, then `dimension`, then `result`.
  2. I added C3 because constraints need to cover all operands and results.
  3. I added C5 because constraints need to cover everything needed for shape inference.